compiler/mir/mirgen.nim

```diff
@@ -0,0 +1,2105 @@
+
+    options: set[GenOption]
+
+const
+  abstractInstTypeClass = abstractInst + tyUserTypeClasses
+  # TODO: this set shouldn't be needed. ``tyUserTypeClass`` and
+  #       ``tyUserTypeClassInst`` should be turned into either aliases or
+  #       ``tyGenericInst`` types when they're resolved
+  ComplexExprs = {nkIfExpr, nkCaseStmt, nkBlockExpr, nkTryStmt}
+    ## The expression that are treated as complex and which are transformed
+    ## into assignments-to-temporaries
+
+func len[T](x: HOslice[T]): int {.inline.} =
+  int(x.b - x.a)
+
+func halfOpen[T](x: Slice[T]): HOslice[T] {.inline.} =
+  ## Constructs a half-open slice from a normal slice. `x` is expected to
+  ## already represent a half-open slice
+  HOslice[T](a: x.a, b: x.b)
+
+template toOpenArray[T, I](x: seq[T], slice: Slice[I]): openArray[T] =
+  toOpenArray(x, slice.a.int, slice.b.int)
+
+template toOpenArray[T, I](x: seq[T], slice: HOslice[I]): openArray[T] =
+  toOpenArray(x, slice.a.int, slice.b.int - 1)
+
+func isHandleLike(t: PType): bool =
+  t.skipTypes(abstractInst).kind in {tyPtr, tyRef, tyLent, tyVar, tyOpenArray}
+
+# XXX: copied from ``injectdestructors``. Move somewhere common
+proc isCursor(n: PNode): bool =
+  ## Computes whether the expression `n` names a location that is a cursor
+  case n.kind
+  of nkSym:
+    sfCursor in n.sym.flags
+  of nkDotExpr:
+    isCursor(n[1])
+  of nkCheckedFieldExpr:
+    isCursor(n[0])
+  else:
+    false
+
+func endsInNoReturn(n: PNode): bool =
+  ## Tests and returns whether the simple or compound statement `n` ends in a
+  ## no-return statement
+
+  # TODO: this is a patched version of ``sem.endsInNoReturn`` that also
+  #       considers ``nkPragmaBlock``. Move this procedure somewhere common and
+  #       replace ``sem.endsInNoReturn`` with it
+  const SkipSet = {nkStmtList, nkStmtListExpr, nkPragmaBlock}
+  var it {.cursor.} = n
+  while it.kind in SkipSet and it.len > 0:
+    it = it.lastSon
+
+  result = it.kind in nkLastBlockStmts or
+    (it.kind in nkCallKinds and it[0].kind == nkSym and
+     sfNoReturn in it[0].sym.flags)
+
+func canonicalExpr(n: PNode): PNode =
+  ## Returns the canonical expression, i.e. the expression without leading
+  ## pragma blocks or empty statement lists
+  func skipped(n: PNode): PNode =
+    case n.kind
+    of nkPragmaBlock:
+      result = n.lastSon
+    of nkStmtListExpr:
+      # see if all nodes except the last one are empty. If they are, the
+      # ``nkStmtListExpr`` is skipped
+      for i in 0..<n.len-1:
+        if n[i].kind != nkEmpty:
+          # there are statements in the list -> the list needs to be kept
+          return n
+
+      result = n.lastSon
+    else:
+      result = n
+
+  var n {.cursor.} = n
+  while (let it = skipped(n); it != n):
+    n = it
+
+  result = n
+
+func selectWhenBranch(n: PNode, isNimvm: bool): PNode =
+  assert n.kind == nkWhen
+  if isNimvm: n[0][1]
+  else:       n[1][0]
+
+func selectDefKind(s: PSym): range[mnkDef..mnkDefCursor] =
```

Comment:
I do want a `range` return type here, both for documentation purposes and so that one doesn't have to use an `else: unreachable()` branch when using the result with a `case`.